src/Build/Utilities/EngineFileUtilities.cs

```diff
@@ -152,8 +219,193 @@ FileMatcher fileMatcher
+                                        XMakeElements.itemGroup);
+                                }
+
+                                break;
+
+                            // Logging mechanism received from Evaluator.
+                            case ILoggingService loggingService:
+                                if (buildEventContext != null)
+                                {
+                                    loggingService.LogWarning(
+                                        buildEventContext,
+                                        string.Empty,
+                                        new BuildEventFileInfo(buildEventFileInfoFullPath),
+                                        DriveEnumeratingWildcardMessageResourceName,
+                                        EscapingUtilities.UnescapeAll(filespecUnescaped),
+                                        XMakeAttributes.include,
+                                        XMakeElements.itemGroup);
+                                }
+
+                                break;
+
+                            // Logging mechanism received from Evaluator and LazyItemEvaluator.IncludeOperation.
+                            case EvaluationLoggingContext evaluationLoggingContext:
+                                if (importLocation != null)
+                                {
+                                    evaluationLoggingContext.LogWarning(
+                                        DriveEnumeratingWildcardMessageResourceName,
+                                        EscapingUtilities.UnescapeAll(filespecUnescaped),
+                                        XMakeAttributes.project,
+                                        XMakeElements.import);
+                                }
+                                else if (excludeFileSpecIsUnusable)
+                                {
+                                    evaluationLoggingContext.LogWarning(
+                                        DriveEnumeratingWildcardMessageResourceName,
+                                        EscapingUtilities.UnescapeAll(fileSpec),
+                                        XMakeAttributes.include,
+                                        XMakeElements.itemGroup);
+                                }
+                                else
+                                {
+                                    evaluationLoggingContext.LogWarning(
+                                        DriveEnumeratingWildcardMessageResourceName,
+                                        EscapingUtilities.UnescapeAll(fileSpec),
+                                        XMakeAttributes.exclude,
+                                        XMakeElements.itemGroup);
+                                }
+
+                                break;
+
+                            default:
+                                throw new NotImplementedException($"Logging type {loggingMechanism.GetType()} " +
```

Comment:
I think this would be better off using our special "it's MSBuild's fault, not the user's fault" type:

```suggestion
                                throw new InternalErrorException($"Logging type {loggingMechanism.GetType()} " +
```